### PR TITLE
[Feature] Proper access to craft via UUID

### PR DESCRIPTION
**Pull request by DerToaster98**
### Describe in detail what your pull request accomplishes
Gives each craft a UUID. Crafts are now tracked in a weak map via their UUID. 
Signs will now get a entry into their persistentdatacontainer that represents the craft UUID.
That way you can find the relevant craft for a sign more reliably (needed for a upcoming PR that will greatly change sign codes and make them easier to expand and more maintainable while also allowing other plugins to override the behavior of a specific sign in a easier way).

### Checklist
- [ ] Proper internationalization (not needed)
- [x] Tested
- [ ] Performance tested (can potentially give performance boosts when many crafts are active)

### CODE REVIEW

File: Movecraft/src/main/java/net/countercraft/movecraft/listener/CraftAssembleListener.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import net.countercraft.movecraft.events.CraftReleaseEvent;
+import net.countercraft.movecraft.util.MathUtils;
+import org.bukkit.block.Block;
+import org.bukkit.block.TileState;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+import org.bukkit.persistence.PersistentDataType;
+import org.jetbrains.annotations.NotNull;
+
+public class CraftAssembleListener implements Listener {
```
Comment (TylerS1066) [inline Movecraft/src/main/java/net/countercraft/movecraft/listener/CraftAssembleListener.java:16]: The naming convention should match the event, `CraftPilotListener`.

File: Movecraft/src/main/java/net/countercraft/movecraft/listener/CraftAssembleListener.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    public void onAssembly(@NotNull CraftPilotEvent event) {
+        // Walk through all signs and set a UUID in there
+        final Craft craft = event.getCraft();
+
+        // Now, find all signs on the craft...
+        for (MovecraftLocation mLoc : craft.getHitBox()) {
+            Block block = mLoc.toBukkit(craft.getWorld()).getBlock();
+            // Only interested in signs, if no sign => continue
+            // TODO: Just limit to signs?
+            if (!(block.getState() instanceof TileState))
```
Comment (TylerS1066) [inline Movecraft/src/main/java/net/countercraft/movecraft/listener/CraftAssembleListener.java:28]: That is a good question, I think for now let's limit ourselves to signs.

File: Movecraft/src/main/java/net/countercraft/movecraft/listener/CraftAssembleListener.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                continue;
+            // Sign located!
+            TileState tile = (TileState) block.getState();
+            // Add the marker
+            tile.getPersistentDataContainer().set(
+                    MathUtils.KEY_CRAFT_UUID,
```
Comment (TylerS1066) [inline Movecraft/src/main/java/net/countercraft/movecraft/listener/CraftAssembleListener.java:34]: Let's move this to the craft class.

File: Movecraft/src/main/java/net/countercraft/movecraft/listener/CraftAssembleListener.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                    PersistentDataType.STRING,
+                    craft.getUUID().toString()
+            );
+            tile.update();
+        }
+    }
+
+    @EventHandler
+    public void onDisassembly(@NotNull CraftReleaseEvent event) {
```
Comment (TylerS1066) [inline Movecraft/src/main/java/net/countercraft/movecraft/listener/CraftAssembleListener.java:43]: Same here, please follow the naming convention of the events.

File: Movecraft/src/main/java/net/countercraft/movecraft/craft/BaseCraft.java [hunk 5]
```diff
@@ -75,6 +72,13 @@ public abstract class BaseCraft implements Craft {
 
     private final CraftDataTagContainer dataTagContainer = new CraftDataTagContainer();
 
+    private final UUID uuid = UUID.randomUUID();
+
+    {
+        // Map to craft
+        Hidden.uuidToCraft.put(uuid, this);
+    }
```
Comment (TylerS1066) [inline Movecraft/src/main/java/net/countercraft/movecraft/craft/BaseCraft.java:80]: This should probably be in the constructor.

**Review by TylerS1066 — CHANGES_REQUESTED** [review]

File: Movecraft/src/main/java/net/countercraft/movecraft/listener/CraftAssembleListener.java (hunk 1 above)
Comment (DerToaster98) [reply Movecraft/src/main/java/net/countercraft/movecraft/listener/CraftAssembleListener.java:16]: For the moment i limited the stuff in the listener to signs, the mark and unmark methods accept the TileState though. That is, if some addon wants to mark, let's say dispensers (i need to do that for my director changes), they can just do that with the given methods.

**Review by TylerS1066 — APPROVED** [review]

**Comment by TylerS1066** [comment]
We've got ourselves some lovely merge conflicts, it's probably best if you rebase this PR.

**Comment by DerToaster98** [comment]
Resolved the conflicts, were just some imports.

**Comment by TylerS1066** [comment]
Looks like we do need imports for `UUID` and some other classes.

**Comment by DerToaster98** [comment]
> Looks like we do need imports for `UUID` and some other classes, probably need to use `import java.util.*;`.

Oh, yes, i overlooked that when resolving it, Fixing it rn...

**Comment by TylerS1066** [comment]
Looks like we need to add the following imports to `Movecraft.java`:
```java
import net.countercraft.movecraft.features.status.StatusManager;
import net.countercraft.movecraft.features.status.StatusSign;
```

**Comment by TylerS1066** [comment]
Was this tested on a server?  It seems to not be operating correctly.
```
java.lang.NullPointerException: Cannot invoke "java.util.UUID.hashCode()" because the return value of "net.countercraft.movecraft.craft.BaseCraft.getUUID()" is null
        at net.countercraft.movecraft.craft.BaseCraft.hashCode(BaseCraft.java:555) ~[Movecraft_42881f2cc80b07b3c7b3c614d5de0ba6e8ab37e2.jar:?]
        at java.util.HashMap.hash(HashMap.java:338) ~[?:?]
        at java.util.HashMap.put(HashMap.java:610) ~[?:?]
        at java.util.HashSet.add(HashSet.java:221) ~[?:?]
        at net.countercraft.movecraft.craft.CraftManager.getPlayerCraftsInWorld(CraftManager.java:334) ~[Movecraft_42881f2cc80b07b3c7b3c614d5de0ba6e8ab37e2.jar:?]
        at net.countercraft.movecraft.features.contacts.ContactsManager.runRecentContacts(ContactsManager.java:124) ~[Movecraft_42881f2cc80b07b3c7b3c614d5de0ba6e8ab37e2.jar:?]
        at net.countercraft.movecraft.features.contacts.ContactsManager.run(ContactsManager.java:35) ~[Movecraft_42881f2cc80b07b3c7b3c614d5de0ba6e8ab37e2.jar:?]
        at org.bukkit.craftbukkit.v1_18_R2.scheduler.CraftTask.run(CraftTask.java:101) ~[purpur-1.18.2.jar:git-Purpur-"dc4a29b"]
        at org.bukkit.craftbukkit.v1_18_R2.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:57) ~[purpur-1.18.2.jar:git-Purpur-"dc4a29b"]
        at com.destroystokyo.paper.ServerSchedulerReportingWrapper.run(ServerSchedulerReportingWrapper.java:22) ~[purpur-1.18.2.jar:git-Purpur-"dc4a29b"]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
        at java.lang.Thread.run(Thread.java:840) ~[?:?]
```

**Comment by TylerS1066** [comment]
Simple enough fix, `Uuid` vs `UUID`.

**Pull request merged**